File: go/death_handler.go
package headgate

import "context"

// DeathReason identifies why a job permanently entered the archive.
type DeathReason string

// Reasons reported to a DeathHandler.
const (
	DeathAttemptsExhausted DeathReason = "attempts_exhausted"
	DeathSkipped           DeathReason = "skipped"
	DeathDeadlineExceeded  DeathReason = "deadline_exceeded"
)

// DeathEvent is emitted only after a fence-verified transition to archived succeeds.
// Envelope returns a deep copy so one callback cannot alter a later callback's view.
type DeathEvent struct {
	envelope Envelope
	reason   DeathReason
	err      string
}

func newDeathEvent(envelope Envelope, reason DeathReason, err string) DeathEvent {
	return DeathEvent{
		envelope: cloneEnqueueBatch([]Envelope{envelope})[0],
		reason:   reason,
		err:      err,
	}
}

// Envelope returns a deep copy of the archived job envelope.
func (e DeathEvent) Envelope() Envelope { return cloneEnqueueBatch([]Envelope{e.envelope})[0] }

// Reason returns why the job was archived.
func (e DeathEvent) Reason() DeathReason { return e.reason }

// ErrorMessage returns the terminal error message, when one was recorded.
func (e DeathEvent) ErrorMessage() string { return e.err }

// TerminalState returns the lifecycle state produced by a death transition.
func (e DeathEvent) TerminalState() string { return "archived" }

// DeathHandler observes a job once when it becomes permanently archived, never once per
// ordinary retry. The durable transition has completed before this method is called.
type DeathHandler interface {
	HandleDeath(context.Context, DeathEvent)
}

// DeathHandlerFunc adapts a function to DeathHandler.
type DeathHandlerFunc func(context.Context, DeathEvent)

// HandleDeath calls f with the archived job event.
func (f DeathHandlerFunc) HandleDeath(ctx context.Context, event DeathEvent) {
	f(ctx, event)
}

func emitDeath(ctx context.Context, handlers []DeathHandler, event DeathEvent) {
	for _, handler := range handlers {
		handler.HandleDeath(ctx, event)
	}
}

File: crates/headgate/src/death_handler.rs
//! Once-only callbacks after a job durably enters the archive.

use headgate_core::Envelope;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum DeathReason {
    AttemptsExhausted,
    Skipped,
    DeadlineExceeded,
}

/// Owned error metadata plus an immutable snapshot of the job that just became archived.
#[derive(Clone, Debug)]
pub struct DeathEvent {
    envelope: Envelope,
    reason: DeathReason,
    error: String,
}

impl DeathEvent {
    pub(crate) fn new(envelope: &Envelope, reason: DeathReason, error: impl Into<String>) -> Self {
        Self {
            envelope: envelope.clone(),
            reason,
            error: error.into(),
        }
    }

    pub fn envelope(&self) -> &Envelope {
        &self.envelope
    }

    pub fn reason(&self) -> DeathReason {
        self.reason
    }

    pub fn error(&self) -> &str {
        &self.error
    }

    pub fn terminal_state(&self) -> &'static str {
        "archived"
    }
}

/// Synchronous notification after the fence-verified archive transition has committed.
pub trait DeathHandler: Send + Sync + 'static {
    fn on_death(&self, event: &DeathEvent);
}

pub struct DeathHandlerFn<F>(F);

impl<F> DeathHandlerFn<F> {
    pub fn new(function: F) -> Self {
        Self(function)
    }
}

impl<F> DeathHandler for DeathHandlerFn<F>
where
    F: Fn(&DeathEvent) + Send + Sync + 'static,
{
    fn on_death(&self, event: &DeathEvent) {
        (self.0)(event);
    }
}

pub(crate) fn emit_death(handlers: &[std::sync::Arc<dyn DeathHandler>], event: DeathEvent) {
    for handler in handlers {
        handler.on_death(&event);
    }
}
